model/scratch/dffml_model_scratch/ridge.py

```diff
@@ -0,0 +1,226 @@
+
+
+    async def fit(self, X, y, sample_weight=None):
+        X = np.asarray(X)
+        y = np.array(y)
+        if len(X.shape) < 2:
+            X = X.reshape(-1, 1)
+        X, y, X_offset, y_offset, X_scale = await self._preprocess_data(
+                X, y, self.fit_intercept, self.normalize, self.copy_X,
+                sample_weight=sample_weight, return_mean=True)
+        params = {}
+        self.coef_, self.n_iter_ = await _ridge_regression(
+                X, y, alpha=self.alpha, sample_weight=sample_weight,
+                max_iter=self.max_iter, tol=self.tol,
+                random_state=self.random_state, return_n_iter=True,
+                return_intercept=False, check_input=False, **params)
+        await self._set_intercept(X_offset, y_offset, X_scale)
+        self.accuracy_value = await self.coeff_of_deter(y, \
+            np.dot(np.asarray(X, dtype=float), self.coef_.T)  + self.intercept_)
+        file_addr = self.parent.config.directory +"\\"+ self._features_hash_
+        with open(file_addr + "weights.txt", 'w') as f:
+            f.write(",".join(map(str, self.coef_)))
+        with open(file_addr + "intercept.txt", 'w') as f:
+            f.write(str(self.intercept_))
+        with open(file_addr + "acc.txt", 'w') as f:
+            f.write(str(self.accuracy_value))
+        return self
+    
+    async def _set_intercept(self, X_offset, y_offset, X_scale):
+        """Set the intercept_
+        """
+        if self.fit_intercept:
+            self.coef_ = self.coef_ / X_scale
+            self.intercept_ = y_offset - np.dot(X_offset, self.coef_.T)
+        else:
+            self.intercept_ = 0.
+
+    async def train(self, sources: Sources):
+        data = []
+        async for repo in sources.with_features(
+            self.features + [self.parent.config.predict.NAME]
+        ):
+            feature_data = repo.features(
+                self.features + [self.parent.config.predict.NAME]
+            )
+            slice_ = [feature_data[data] for data in self.features]
+            data.append(slice_)
+            self.yData = np.append(
+                self.yData, feature_data[self.parent.config.predict.NAME]
+            )
+        self.xData = np.asarray(data, dtype=float).reshape(-1, len(self.features))
+        await self.best_fit_line()
+
+    async def accuracy(self, sources: Sources) -> Accuracy:
+        file_addr = self.parent.config.directory +"\\"+ self._features_hash_
+        try:
+            with open(file_addr + "acc.txt", "r") as f:
+                self.accuracy_value = float(f.read())
+        except:
+            raise ModelNotTrained("Train model before assessing for accuracy.")
+        return Accuracy(self.accuracy_value)
+
+    async def predict(
+        self, repos: AsyncIterator[Repo]
+    ) -> AsyncIterator[Tuple[Repo, Any, float]]:
+        try:
+            file_addr = self.parent.config.directory +"\\"+ self._features_hash_

```

Comment:
Let's make sure to use `pathlib.Path` or `os.path.join` to join elements in a path. This way we maintain unix and windows support.

Reply:
For example:

```python
file_addr = pathlib.Path(self.parent.config.directory) / self._features_hash_
```

Reply:
You'd then want

```python
with open(str(file_addr) + "weights.txt", 'r') as f:
```

Reply:
done